### PR TITLE
Add v1beta1 version of CustomRun CRD

### PR DESCRIPTION
**Pull request by vsinghai**
# Changes

This commit adds a v1beta1 version of the CustomRun CRD, and support to the webhook.

Fixes issue [#5154](https://github.com/tektoncd/pipeline/issues/5154)
Fixes issue [#5153](https://github.com/tektoncd/pipeline/issues/5153)
Fixes issue [#5532](https://github.com/tektoncd/pipeline/issues/5532)

/cc @jerop
/kind misc

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- n/a Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- n/a Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by vsinghai** [comment]
do I need `conversions`, like stated [here](https://github.com/tektoncd/pipeline/blob/df7abd26b9800c767fb49897aaaf945dd76ced3c/config/300-pipeline.yaml#L70-L77), for v1alpha1.Runs to v1beta1.CustomRuns?

**Comment by XinruZhang** [comment]
> do I need `conversions`, like stated [here](https://github.com/tektoncd/pipeline/blob/df7abd26b9800c767fb49897aaaf945dd76ced3c/config/300-pipeline.yaml#L70-L77), for v1alpha1.Runs to v1beta1.CustomRuns?

From the API compatibility policy perspective, it's okay that we don't convert it (per `Rule #2` in [K8s API Deprecation Policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/)) (meaning we deprecate the entire alpha track of the Custom Task Run).

The reason that PipelineRun needs that conversion is we want to support both v1beta1 and v1 version at the same time.

**Comment by lbernick** [comment]
> > do I need `conversions`, like stated [here](https://github.com/tektoncd/pipeline/blob/df7abd26b9800c767fb49897aaaf945dd76ced3c/config/300-pipeline.yaml#L70-L77), for v1alpha1.Runs to v1beta1.CustomRuns?
> 
> From the API compatibility policy perspective, it's okay that we don't convert it (per `Rule #2` in [K8s API Deprecation Policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/)) (meaning we deprecate the entire alpha track of the Custom Task Run).
> 
> The reason that PipelineRun needs that conversion is we want to support both v1beta1 and v1 version at the same time.

I think we would ideally like to have conversion here because it will allow us to support v1alpha1 Runs and v1beta1 CustomRuns at the same time. Technically it's not necessary since v1alpha1 Run is in alpha but it feels a bit user hostile to drop support for Runs at the same time we're adding support for CustomRuns. It will force people to copy all their Run definitions somewhere, delete them, upgrade tekton pipelines, and then recreate their old run definitions as CustomRuns. (This would also be the case if we rename the "Kind" of the existing CRD to "CustomRun".)

Unfortunately I'm not sure if there's an easy way to rename a CRD/implement conversion between two different "Kinds". One thing we could try is to add a v1beta1 version of Run that deserializes to the CustomRun go struct, in addition to adding a CustomRun CRD with a v1beta1 version, and probably deprecate the Run CRD after giving people a release or two to convert to the new CRD. (I'm not totally sure if this would work)

Another idea would be to have the new CRD and update the pipelinerun reconciler code to allow reconciling both v1alpha1 Runs and v1beta1 custom runs, and not implement conversion. We would probably want to remove support for v1alpha1 runs after a release or two. I'm fairly sure this would work.

happy to help here but want to hear others' thoughts first -- @tektoncd/core-maintainers is this making sense/do you know how to rename a CRD?

**Comment by abayer** [comment]
I've never actually seen anyone rename a CRD - I'm pretty sure you can't convert between two different `Kind`s, yeah. Some research is definitely needed here.

**Comment by XinruZhang** [comment]
I would say the main focus is how do we want our users to conceive the "alpha" feature.

IMHO alpha version track is designed to serve as experimental features, meaning we really don't want our users to rely on alpha features in their prod env. They are only suppose to use it for experiments and they should expect the alpha features can be different, backward incompatible across different releases.

However, IIUC, this "alpha" feature has existed for a long time and some users have already heavily relied on it (have they?), we can't ignore this real life situation, in this case, I strongly recommend we need to support both for a while, giving them enough buffer time to migrate.

But in order to form a formal standard around **stability** of each version track, I would suggest it might be better if we explicitly tell users that **supporting both alpha and beta versions this time is an exception**, this decision is not based on the guidance we'd like to follow, but based on the reality that we haven't actively promoting alpha features to beta in the past. While we are planning to do it in the future.

**Comment by vsinghai** [comment]
Thanks for the great conversation @lbernick @XinruZhang @abayer . I would like to ask a couple of questions regarding this situation. To my understanding, `v1alpha1.Runs` and `v1beta1.CustomRuns`, in theory, should be mutually exclusive. Therefore, I am not clear as to why we should add the conversion, when we should rather push the user to start using `CustomRuns` while still providing support for `Runs` (for a limited amount of time). In the end, the whole point of the K8 version control is to deprecate the less stable versions, i.e. `v1alpha1` (where `Runs` is stored), and keep upgrading to stable versions. If we are trying to upgrade `Runs` to `CustomRuns` from `v1alpha` to `v1beta1`, but still allow the users to be able to use the `v1alpha1` feature, doesn't that negate the whole point of constantly trying to upgrade our K8 versioning? I understand @lbernick point that it would be harsh to the user to suddenly drop `Runs` out of the blue and force everyone to update their code, but would a slow push of `CustomRuns` paired with telling users that `Runs` will no longer be supported help solve that issue?

Just providing some food for thought here, let me know what you think!

**Comment by lbernick** [comment]
> Thanks for the great conversation @lbernick @XinruZhang @abayer . I would like to ask a couple of questions regarding this situation. To my understanding, `v1alpha1.Runs` and `v1beta1.CustomRuns`, in theory, should be mutually exclusive. Therefore, I am not clear as to why we should add the conversion, when we should rather push the user to start using `CustomRuns` while still providing support for `Runs` (for a limited amount of time).

The main purpose of implementing conversion is so that we only have to support one golang version in our reconciler code, because each served version (version that users can use) gets converted into the same stored version (version stored in etcd and used by our reconciler code). If we aren't able to implement conversion, our reconciler code will have to support both v1alpha1 Run and v1beta1 CustomRun. Contrast this to PipelineRun: our stored version is v1beta1, but once we start serving v1, it will be converted into a v1beta1 PipelineRun for use in our reconciler, so the reconciler code only needs to handle one version at a time.

I agree we'd like to ask users to swap to customruns while still supporting runs for a short window.

> In the end, the whole point of the K8 version control is to deprecate the less stable versions, i.e. `v1alpha1` (where `Runs` is stored), and keep upgrading to stable versions. If we are trying to upgrade `Runs` to `CustomRuns` from `v1alpha` to `v1beta1`, but still allow the users to be able to use the `v1alpha1` feature, doesn't that negate the whole point of constantly trying to upgrade our K8 versioning? I understand @lbernick point that it would be harsh to the user to suddenly drop `Runs` out of the blue and force everyone to update their code, but would a slow push of `CustomRuns` paired with telling users that `Runs` will no longer be supported help solve that issue?

I agree we'd like to remove support for Runs (I'm not sure exactly what you mean by a "slow push" here). The question is just how best to create a release where we support both Runs and CustomRuns, before creating a release where we remove support for Runs.

**Comment by vsinghai** [comment]
/cc @dibyom 

please review

**Review by jerop — APPROVED** [review]
There's no conversion in this PR, it's just adding a new CRD instead of a new version, so I think the conversion conversation should not be a blocker here - we can move it to https://github.com/tektoncd/pipeline/pull/5470

(I believe we don't need conversion because this is a new CRD instead of a new version)

**Comment by lbernick** [comment]
I'm not sure this PR should go in yet-- we're exposing a CRD that doesn't actually do anything when you create it

File: config/300-customrun.yaml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    app.kubernetes.io/part-of: tekton-pipelines
+    pipeline.tekton.dev/release: "devel"
+    version: "devel"
+spec:
+  group: tekton.dev
+  preserveUnknownFields: false
+  versions:
+  - name: v1beta1
+    served: true
+    storage: true
```
Comment (jerop) [inline config/300-customrun.yaml:30]: @vsinghai can we set both of these fields to false?

Comment (vsinghai) [reply config/300-customrun.yaml]: done

**Comment by vsinghai** [comment]
Here is a [doc](https://docs.google.com/document/d/1sJGHtrPq8R2FuIpbUTjWUxQHSssVArapJuFwRC1FD04/edit#) with the roadmap for this PR. The doc summarizes the lengthy conversation we have had in this PR and what the team has agreed upon for next steps for the PR. 

Once again, thank you @lbernick @XinruZhang and @abayer for the great discussion!

**Pull request closed without merging**